src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -138,45 +138,69 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
             ref BoundExpression? receiverOpt,
             ref ImmutableArray<BoundExpression> arguments,
             ref ImmutableArray<RefKind> argumentRefKindsOpt,
-            ref bool invokedAsExtensionMethod,
+            bool invokedAsExtensionMethod,
             Location? interceptableLocation)
         {
             // PROTOTYPE(ic):
             // Add assertions for the possible shapes of calls which could come through this method.
             // When the BoundCall shape changes in the future, force developer to decide what to do here.
 
-            // PROTOTYPE: perhaps a 'TryGet' pattern is more suitable here.
-            if (this._compilation.GetInterceptor(interceptableLocation) is not var (interceptsLocationAttributeData, interceptor))
+            if (this._compilation.TryGetInterceptor(interceptableLocation, _diagnostics) is not var (attributeLocation, interceptor))
             {
                 // The call was not intercepted.
                 return;
             }
 
             Debug.Assert(interceptableLocation != null);
+            Debug.Assert(interceptor.Arity == 0);
+
             if (!method.IsInterceptable)
             {
-                // PROTOTYPE(ic): it was speculated that we could avoid work if we know the current method is not interceptable.
-                // i.e. use this as an early out before even calling Compilation.GetInterceptor.
-                // But by calling 'GetInterceptor' before this, we don't really avoid that work. Is that fine?
-                // PROTOTYPE(ic): eventually we probably want this to be an error but for now it's convenient to just warn
-                // so we can experiment with intercepting APIs that haven't yet been marked.
-                this._diagnostics.Add(ErrorCode.WRN_CallNotInterceptable, interceptsLocationAttributeData.AttributeLocation, method);
+                // PROTOTYPE(ic): Eventually we may want this to be an error.
+                // For now it's convenient to just warn so we can experiment with intercepting APIs that haven't yet been marked.
+                this._diagnostics.Add(ErrorCode.WRN_CallNotInterceptable, attributeLocation, method);
             }
 
-            Debug.Assert(interceptor.Arity == 0);
+            var containingMethod = this._factory.CurrentFunction;
+            Debug.Assert(containingMethod is not null);
+            var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+            if (!AccessCheck.IsSymbolAccessible(interceptor, containingMethod.ContainingType, ref discardedUseSiteInfo))
```

Comment:
>discardedUseSiteInfo [](http://example.com/codeflow?start=98&length=20)

Why ignore use-site diagnostics here? (It's not clear to me what use-site diagnostics might be generated, but is there a reason to ignore them?)

Reply:
I don't see how they could be reported in the feature as currently designed, but if we allowed interceptors to be generic and defined how their type arguments are obtained in the future, it could become relevant to make sure the use site diagnostics are captured.

Might as well just do it right now.